### ISSUE TITLE
Jul2021: Cannot create filter functions

### ISSUE THREAD

**Issue by swingbit** [MonetDB/MonetDB#7116]
**Describe the bug**
When I try to create a custom filter function, I get:
```
sql>CREATE OR REPLACE FILTER FUNCTION maxlevenshtein(s1 string, s2 string, k tinyint) EXTERNAL NAME spinque."maxlevenshtein";
CREATE FILTER FUNCTION: there's a function with the name 'maxlevenshtein' and the same parameters, which causes ambiguous calls
```

However, no such function was created before:
```
sql>select * from sys.functions where name = 'maxlevenshtein';
+----+------+------+-----+----------+------+-------------+--------+--------+-----------+--------+-----------+
| id | name | func | mod | language | type | side_effect | varres | vararg | schema_id | system | semantics |
+====+======+======+=====+==========+======+=============+========+========+===========+========+===========+
+----+------+------+-----+----------+------+-------------+--------+--------+-----------+--------+-----------+
0 tuples
```

No issue with normal functions.

**Software versions**
- MonetDB 5 server 11.41.0 (hg id: f10cd9f0c1)
- OS and version: Fedora 33
- compiled from sources

**Comment by PedroTadim** [comment]
I just fixed this issue on my latest commit, so I'm going to close it. If you still find an issue, please re-open it.

**Comment by swingbit** [comment]
Hi Pedro,
Thanks for your fix.

I am not allowed to reopen this, but I still have an issue:
```
sql>CREATE OR REPLACE FUNCTION contains(s1 string, s2 string, casesensitive boolean) RETURNS boolean EXTERNAL NAME spinque."contains";
operation successful

sql>CREATE OR REPLACE FILTER FUNCTION contains(s1 string, s2 string, casesensitive boolean) EXTERNAL NAME spinque."contains";
CREATE FILTER FUNCTION: there's a function with the name 'contains' and the same parameters, which causes ambiguous calls
```

This time, it is true that there is a function with the same name and the same parameters. But the first is a function, the second is a filter function.

I had a look at how the code changed from Oct2020, and I am afraid that this was the intended behaviour indeed.
If that is the case, may I kindly suggest to reconsider this choice? 

If I understand well, filter functions used to be distinguished earlier on. Now the binding happens at the same level as it does for normal functions. Certainly this makes the code more elegant, I agree on that.

However, this removes an important functionality and has severe implications on existing applications and databases (besides, how can one even upgrade from Oct2020 where such two functions already exist?). The fact that you can use the same function signature for both a selection and a join is a very desirable feature. Having to change their name only for an implementation detail doesn't seem a great way to go. 

Could the code improvement be kept, while saving this functionality? I'm sure I am overlooking something, but I wonder how the situation of an ambiguous call may even arise. 
If you have a filter function, why would you ever prefer to use a cross product + selection instead, or a theta-join? Wouldn't you always prefer to use the filter function if available? If the corresponding filter function isn't always a better choice then it shouldn't be there in the first place.

**Comment by PedroTadim** [comment]
Hi Roberto,

I agree with your side as well. We can leave the ambiguity, but we have to document it somewhere.

In the future we are going to improve the SQL to MAL translation for filter functions. At the moment, we don't verify if the corresponding MAL select and join calls are available for a filter function.

**Comment by PedroTadim** [comment]
I just relaxed the restriction here: https://dev.monetdb.org/hg/MonetDB/rev/a07cecd60acf

**Comment by swingbit** [comment]
Super, just tested. 

Just to be sure I understand well: 

In my quick test I was allowed indeed to use the same SQL signature for a standard function and a filter function.

When using it like this: 
```
explain select * from (values ('abc'),('bcd')) as a(i), (values ('bc'),('ab')) as b(i) where [a.i] contains [b.i, true];
```
it did prefer the filter function. Great.

When I used it like this:
```
explain select * from (values ('abc'),('bcd')) as a(i), (values ('bc'),('ab')) as b(i) where contains(a.i, b.i, true);
```
it used crossproduct+select, even though it is equivalent to the previous one. Not an issue for me, I'm just curious whether this is what you expected.

Two additional questions if I may:
- Is the choice in the first query "stable", i.e. will it always prefer the filter function? I thought the check that you now removed at creation time was due to this choice not being deterministic, but probably I misunderstood that part.
- Is eventually the plan that the second query should be equivalent to the first one and boil down the same function bindings?

Thanks Pedro!

**Comment by PedroTadim** [comment]
I think in the second example you are hitting a change we did in the parser to speed up query compilation about a month ago. Before at parsing, we would test if a identifier existed as a filter function, then generate a filter token (in bison) if that was the case. However the test for the filter function existence for every identifier was expensive. The reason was because it would check the catalog and hit a mutex. As filter functions are not widely used, we removed this check in favor to speed up query compilation.

At the moment, the first query is the current way to express a filter function and I think that should be the way. Sometimes a user may want to use the non filter version explicitly. So my answer to your questions is there must be a way to explicitly call the filter and non filter versions, and you just showed how to do it.
